# INTEGRATED PROJECT PART III-WEAVING THE DATA THREADS OF MAJI NDOGO'S NARRATIVE: A Data-Driven Journey
## Unveiling the Water Crisis in Maji Ndogo

## 1. Introduction


In this notebook, we investigate the water crisis in Maji Ndogo using SQL-based data analysis. Our goal is to identify problems in water access,
understand the geographic and demographic distribution of resources, and recommend a repair prioritization strategy.

First, let's load our sample database:

In [1]:
# Load and activate the SQL extension to allow us to execute SQL in a Jupyter notebook.
%load_ext sql

In [2]:
# Establish a connection to the local database using the '%sql' magic command.
# Replace 'password' with our connection password and `db_name` with our database name. 
# If you get an error here, please make sure the database name or password is correct.

%sql mysql+pymysql://root:*******@LOCALHOST:3306/md_water_services

'Connected: root@md_water_services'

### 2. GETTING USED TO THE DATABASE
**Slides: 1–4**

We need to understand the database structure. we need an ERD 

### md_water_services Entity Relationship Diagram ERD
![ERD](md_water_services_ERD.png)

Note that the `visits` table is the central table. `location_id`, `source_id` and `assigned_employee_id` are **primary keys** in their respective tables, but are **all foreign keys in visits**. These are mostly one-to-many relationships.

The `visits` table logs all the times we've been to different places, and we can see that **some locations have been visited multiple times.** On the
other hand, the `location `table has all the specifics about each place we've been **but it only includes each location once.** So, it's a `one-to-many
relationship:` for `each unique location in the location table, there might be many corresponding records in the visits table detailing all the
different times we went there.`

## 3. The relationship between the `visits` and `water_quality` tables.
**Slides: 5**

In [3]:
%%sql
select*
from visits
limit 3;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
3 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40)]

In [4]:
%%sql
select*
from water_quality
limit 3;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
3 rows affected.


[(0, 0, 1), (1, 1, 1), (2, 5, 1)]

For **every entry in the `visits` table**, there should be **one unique
corresponding record in the water_quality table**. This means each visit recorded is associated with a specific water quality score, ensuring a
**one-to-one relationship** between the `visits` and `water_quality` tables.

##**But if we look at the ERD, it shows a `many-to-one` relationship. This does not agree with our thinking. Errors like these can cause problems, so let's
fix that.**

## 4. Changing the cardinality of the visits and water_quality tables

We should normally be careful making a change like this. 
- First you should check that `record_id` is unique for both tables, and are indeed one-to-one.
- right-click on the relationship line and select **Edit relationship**
-  then at the bottom, select the **Foreign key tab**
-   change the Cardinality to **one-to-one**.
-    Now the relationship line should indicate a **one-to-one **relation.

In [5]:
%sql SELECT * FROM visits limit 3;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
3 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40)]

In [6]:
%sql SELECT * FROM water_quality limit 3;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
3 rows affected.


[(0, 0, 1), (1, 1, 1), (2, 5, 1)]

## 5. Integrating the Auditor's report
**slide 6**

1. create the table `auditor_report` first

In [7]:
%%sql

DROP TABLE IF EXISTS `auditor_report`;
CREATE TABLE `auditor_report` (
`location_id` VARCHAR(32),
`type_of_water_source` VARCHAR(64),
`true_water_source_score` int DEFAULT NULL,
`statements` VARCHAR(255)
);

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
0 rows affected.
0 rows affected.


[]

2. Select the database
3. right-click on it
4. choose Table Data Import Wizard. 
5. All settings should be left on default

In [8]:
%%sql
SHOW COLUMNS
FROM auditor_report;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
4 rows affected.


[('location_id', 'varchar(32)', 'YES', '', None, ''),
 ('type_of_water_source', 'varchar(64)', 'YES', '', None, ''),
 ('true_water_source_score', 'int', 'YES', '', None, ''),
 ('statements', 'varchar(255)', 'YES', '', None, '')]

In [9]:
%sql SELECT * FROM auditor_report limit 3;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
3 rows affected.


[('SoRu34980', 'well', 1, "Residents admired the official's commitment to enhancing urban life, praising their cooperative and inclusive approach."),
 ('AkRu08112', 'well', 3, "Villagers spoke highly of the official's dedication and genuine interest in their lives, fostering a sense of belonging and appreciation."),
 ('AkLu02044', 'river', 0, "Villagers were touched by the official's interactions, noting their humility, strong work ethic, and respectful attitude.")]

## 6. Comparison Of auditor_report and our dataset
**slide 7-13**

## Questions to be answered

1. Is there a difference in the scores?
2. If so, are there patterns?

## 6.1. ***QUESTION #1*** Is there a difference in the scores?

Compare the quality scores in the `water_quality` table to the `auditor's scores`. The `auditor_report` table
used `location_id`, but the quality scores table only has a `record_id `we can use. The `visits` table links `location_id` and `record_id`, so we
can link the `auditor_report` table and `water_quality` using the `visits` table.

In [10]:
%%sql
SELECT
    * 
FROM 
    water_quality 
limit 3;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
3 rows affected.


[(0, 0, 1), (1, 1, 1), (2, 5, 1)]

In [24]:
%%sql
SELECT 
    location_id, 
    true_water_source_score 
FROM 
    auditor_report 
order by true_water_source_score desc 
limit 3;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
3 rows affected.


[('KiRu30351', 9), ('AmRu13718', 9), ('KiHa22748', 9)]

In [12]:
%%sql
SELECT
    * 
FROM 
    visits 
limit 3;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
3 rows affected.


[(0, 'SoIl32582', 'SoIl32582224', datetime.datetime(2021, 1, 1, 9, 10), 1, 15, 12),
 (1, 'KiRu28935', 'KiRu28935224', datetime.datetime(2021, 1, 1, 9, 17), 1, 0, 46),
 (2, 'HaRu19752', 'HaRu19752224', datetime.datetime(2021, 1, 1, 9, 36), 1, 62, 40)]

## 6.1.1 Join the `visits` table to the `auditor_report` table.
You will need 
- `location_id` from `visits` and `auditor_report` tables
- `true_water_source_score` from `auditor_report`table
- `record_id` from `visits` table
  

In [13]:
%%sql
SELECT
    auditor_report.location_id AS audit_location,
    auditor_report.true_water_source_score,
    visits.location_id AS visit_location,
    visits.record_id
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
limit 5;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
5 rows affected.


[('SoRu34980', 1, 'SoRu34980', 5185),
 ('AkRu08112', 3, 'AkRu08112', 59367),
 ('AkLu02044', 0, 'AkLu02044', 37379),
 ('AkHa00421', 3, 'AkHa00421', 51627),
 ('SoRu35221', 0, 'SoRu35221', 28758)]

## 6.1.2 Retrieve the corresponding scores from the `water_quality` table.
JOIN the `visits` table and the `water_quality` table, using the
`record_id` as the connecting key.

You will need 
- subjective_quality_score
- record_id 

In [14]:
%%sql
SELECT
    auditor_report.location_id AS audit_location,
    auditor_report.true_water_source_score,
    visits.location_id AS visit_location,
    visits.record_id,
    water_quality.subjective_quality_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id
limit 5;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
5 rows affected.


[('SoRu34980', 1, 'SoRu34980', 5185, 1),
 ('AkRu08112', 3, 'AkRu08112', 59367, 3),
 ('AkLu02044', 0, 'AkLu02044', 37379, 0),
 ('AkHa00421', 3, 'AkHa00421', 51627, 3),
 ('SoRu35221', 0, 'SoRu35221', 28758, 0)]


 
 ## 6.1.3 The ***Audits*** location and the ***visit location*** is the same, so we DROP one of the them.
 - Leave `record_id` and rename the scores to `surveyor_score` and `auditor_score` to make it clear which scores
we're looking at in the results set.

In [16]:
%%sql
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id
limit 5;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
5 rows affected.


[('SoRu34980', 5185, 1, 1),
 ('AkRu08112', 59367, 3, 3),
 ('AkLu02044', 37379, 0, 0),
 ('AkHa00421', 51627, 3, 3),
 ('SoRu35221', 28758, 0, 0)]

### Merged record count

In [18]:
%%sql

WITH scores as (
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id)
-- limit 5
SELECT 
    COUNT(DISTINCT record_id)
FROM
    scores;


 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
1 rows affected.


[(2698,)]

## 6.1.4 ***So is there a difference between the auditor scores and the surveyor scores***
- We can have a `WHERE` clause and check if `surveyor_score = auditor_score` or
- we can subtract the two scores and check if the result is 0.

In [19]:
%%sql
with scores as(
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id
)
SELECT *
FROM scores
WHERE auditor_score<>surveyor_score
limit 5;

 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
5 rows affected.


[('AkRu05215', 21160, 3, 10),
 ('KiRu29290', 7938, 3, 10),
 ('KiRu29290', 8076, 3, 10),
 ('KiRu29290', 8108, 3, 10),
 ('KiRu29290', 8168, 3, 10)]

In [20]:
%%sql
with scores as (
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id
WHERE true_water_source_score<>subjective_quality_score
    )
SELECT 
    COUNT(distinct record_id)
FROM scores;


 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
1 rows affected.


[(193,)]

## - ***193*** Rows returned for differing scores


# ** ***OR for similar rows/matching site scores***

In [21]:
%%sql
with scores as (
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id
WHERE true_water_source_score=subjective_quality_score
    )
SELECT 
    COUNT(distinct record_id)
FROM scores;


 * mysql+pymysql://root:***@LOCALHOST:3306/md_water_services
1 rows affected.


[(2505,)]

## **** So ***YES*** -the scores are different ****

## 6.1.5 Fix multiple location visit times

 **If you substitute `<>` with  `=` you get ***2505*** Rows returned**
 - Some of the locations were visited multiple times, so these records are duplicated here. 
 - To fix it, we set `visits.visit_count= 1` in the `WHERE` clause.
 - Make sure you reference the alias you used for `visits` in the join. 

In [ ]:
%%sql
select * from visits 
where location_id= 'KiRu26190'
order by visit_count asc limit	10;

In [ ]:
%%sql
with scores as (
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id
WHERE true_water_source_score=subjective_quality_score
    AND visits.visit_count= 1)
-- limit 5;
SELECT 
    COUNT(distinct record_id)
FROM scores;

In [ ]:
%%sql
with scores as (
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1)
-- limit 5;
SELECT 
    COUNT(distinct record_id)
FROM scores;

-  This results to ***1518 correct*** sites 

## How many sites did the auditor visit ?


In [ ]:
%%sql
SELECT 
    COUNT(*)
FROM 
  auditor_report;  

 -  ***ROUND(1518/1620 x 100) = 94%*** of the records the auditor checked were `correct!!  on the first visit`****
 -  ***ROUND((1620-1518)/1620 x 100) = 6%*** of the records `(102)` the auditor checked  `mismatched!! on the first visit` ****

## 6.1.6 Deep dig into the 102 differing site scores
- work with previous query
- change `visits.visit_count= 1` to `visits.visit_count<>1`

In [ ]:
%%sql

SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1
order by auditor_score desc
limit 5




In [ ]:
%%sql
WITH scores as (
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1
order by auditor_score desc)
-- limit 5

SELECT 
    COUNT(distinct record_id)
FROM scores;
    

In [ ]:
%%sql
WITH scores as (
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id
WHERE true_water_source_score=subjective_quality_score
    AND visits.visit_count<> 1
order by auditor_score desc)
-- limit 5

SELECT 
    COUNT(distinct record_id)
FROM scores;
    



***987*** rows- `not necessarily 987 unique sites`. 

Some locations had:
- Multiple matching visits `after the first visit` 
- Matching and mismatching visits both `after the first visit`

## ***QUESTION #2*** If so, are there patterns?

1. Grab the `type_of_water_source` column from the `water_source` table and call it `survey_source`
2. Use the `source_id` column to JOIN. 
3. Also select the `type_of_water_source` from the `auditor_report` table, and call it `auditor_source`.

In [ ]:
%%sql

SELECT
    auditor_report.location_id AS location_Id,
    auditor_report.type_of_water_source as auditor_source,
    water_source.type_of_water_source as survey_source,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN 
    water_quality 
ON visits.record_id=water_quality.record_Id
JOIN
    water_source
ON water_source.source_id= visits.source_Id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1

limit 5;



## 6.2.1 ***ARE THE SOURCES THE SAME?*** ****

In [ ]:
%%sql

SELECT
    auditor_report.location_id AS location_Id,
    auditor_report.type_of_water_source as auditor_source,
    water_source.type_of_water_source as survey_source,
    visits.record_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN 
    water_quality 
ON visits.record_id=water_quality.record_Id
JOIN
    water_source
ON water_source.source_id= visits.source_Id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1
    AND water_source.type_of_water_source<>auditor_report.type_of_water_source

limit 5;


*** **ALL SOURCES ARE THE SAME** ***

# 6.2.2. Linking records to employees
**SLIDE 14- 17**

Let's look at where these errors may have come from. At some of the locations, `employees assigned scores incorrectly`, and those records
ended up in this results set.

Reasons HOW this can happen.
1. These workers are all humans and make mistakes so this is expected.
2. Unfortunately, the alternative is that someone assigned scores incorrectly on purpose!

- So JOIN the `assigned_employee_id` for all the people on our list from the `visits`
table to the query.
- Remember, the query shows the shows the 102 incorrect records, so when we join the employee data, we can see `which
employees made these incorrect records.`

In [ ]:
%%sql

SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    visits.assigned_employee_id,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1


limit 5;


## 6.2.3 Identify their names from the employees Table
We have employees' names stored along with their IDs, so let's fetch their `names` from the `employees` table instead of the ID's.

In [ ]:
%%sql

SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    employee.employee_name,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id

JOIN employee
ON  employee.assigned_employee_id=visits.assigned_employee_id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1

limit 5;


# SAVE THE QUERY AS A CTE 
- Call it something like `Incorrect_records`.

In [ ]:
%%sql
 WITH incorrect_records as ( 
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    employee.employee_name,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id

JOIN employee
ON  employee.assigned_employee_id=visits.assigned_employee_id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1)

SELECT * 
FROM incorrect_records

limit 5;


## 6.2.4 COUNT THE UNIQUE EMPLOYEES

In [ ]:
%%sql
WITH incorrect_records as ( 
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    employee.employee_name,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id

JOIN employee
ON  employee.assigned_employee_id=visits.assigned_employee_id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1)
SELECT 
    COUNT( DISTINCT employee_name)
FROM
    Incorrect_records;

## 6.2.5 Calculate how many mistakes each employee made.
- Count how many times their name is in Incorrect_records list
- Group them by `name`

In [ ]:
%%sql
WITH incorrect_records as ( 
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    employee.employee_name,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id

JOIN employee
ON  employee.assigned_employee_id=visits.assigned_employee_id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1)

SELECT 
    employee_name,
    COUNT( employee_name) AS number_of_mistakes
FROM
    Incorrect_records
GROUP BY employee_name
ORDER BY number_of_mistakes desc 
limit 17;

**Some of our surveyors are making a lot of "mistakes" while many of the other surveyors are only making a few**

#  ***Gathering some evidence***

**How would we go about finding out if any of our employees are corrupt?***
Let's `say all employees make mistakes`, if someone is `corrupt`, they will be making a lot of "mistakes", `more than average`, for example. But someone
could just be clumsy, so we should try to get more evidence...

- Our auditor noted some of the things he heard on the streets which were quite shady
- He recorded this in the statements column. 
- Considering both of these sources should give us a pretty reliable answer.


## 6.2.6. Find all of the employees who have an `above-average number of mistakes`.

## First calculate the number of times someone's name comes up
1. Call it `error_count`.
2. Calculate the `average` number of mistakes employees made
3. Call that result `avg_error_count_per_empl`, which would be a scalar value

In [ ]:
%%sql
with error_count as (
WITH incorrect_records as ( 
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    employee.employee_name,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id

JOIN employee
ON  employee.assigned_employee_id=visits.assigned_employee_id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1)

SELECT 
    employee_name,
    COUNT( employee_name) AS number_of_mistakes
FROM
    Incorrect_records
GROUP BY employee_name
ORDER BY number_of_mistakes desc )

SELECT
    AVG(number_of_mistakes) AS avg_error_count_per_empl
FROM
    error_count 
limit 5;

## Compare each `employee's error_count` with `avg_error_count_per_empl`.
4. Call this results set our `suspect_list`
5. Remember that we can't use an aggregate result in WHERE, so we have to use `avg_error_count_per_empl` as a subquery

In [ ]:
%%sql -- this is a cte 
with error_count as (
WITH incorrect_records as ( 
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    employee.employee_name,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id

JOIN employee
ON  employee.assigned_employee_id=visits.assigned_employee_id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1)

SELECT 
    employee_name,
    COUNT( employee_name) AS number_of_mistakes
FROM
    Incorrect_records
GROUP BY employee_name
ORDER BY number_of_mistakes desc )

SELECT
    employee_name,
  number_of_mistakes
FROM
    error_count
WHERE 
    number_of_mistakes > (SELECT
                                AVG(number_of_mistakes) AS avg_error_count_per_empl
                            FROM
                                error_count)
limit 5;

## #Optionally CONVERT `incorrect_records` FROM A CTE TO A VIEW
- Replace `WITH` with `CREATE VIEW` like this, and  add the `statements` column  from the `auditor_report` to this table

In [ ]:
%%sql
DROP VIEW IF EXISTS Incorrect_records;
CREATE VIEW Incorrect_records AS (
SELECT
    auditor_report.location_id,
    visits.record_id,
    employee.employee_name,
    auditor_report.true_water_source_score AS auditor_score,
    wq.subjective_quality_score AS surveyor_score,
    auditor_report.statements AS statements
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality AS wq
ON visits.record_id = wq.record_id
JOIN
    employee
ON employee.assigned_employee_id = visits.assigned_employee_id
WHERE
    visits.visit_count =1
    AND auditor_report.true_water_source_score != wq.subjective_quality_score );

In [ ]:
%%sql
SELECT * FROM Incorrect_records
limit 5;

In [ ]:
%%sql  ## count distinct employees using the view
SELECT 
    COUNT( DISTINCT employee_name)
FROM
    Incorrect_records;

In [ ]:
%%sql ## each employees mistakes
SELECT 
    employee_name,
    COUNT( employee_name) AS number_of_mistakes
FROM
    Incorrect_records
GROUP BY employee_name
ORDER BY number_of_mistakes desc 
limit 5;

## #Optionally CONVERT `error_count` FROM A CTE TO A VIEW

In [ ]:
%%sql
DROP VIEW IF EXISTS error_count;
CREATE VIEW error_count as (
WITH incorrect_records as ( 
SELECT
    auditor_report.location_id AS location_Id,
    visits.record_id,
    employee.employee_name,
    auditor_report.true_water_source_score as auditor_score,
    water_quality.subjective_quality_score as surveyor_score,
    auditor_report.statements AS statements
FROM
    auditor_report
JOIN
    visits
ON auditor_report.location_id = visits.location_id
JOIN
    water_quality 
ON visits.record_id=water_quality.record_Id

JOIN employee
ON  employee.assigned_employee_id=visits.assigned_employee_id
WHERE true_water_source_score<>subjective_quality_score
    AND visits.visit_count= 1)

SELECT 
    employee_name,
    COUNT( employee_name) AS number_of_mistakes
FROM
    Incorrect_records
GROUP BY employee_name
ORDER BY number_of_mistakes desc )

### ***avg_error_count_per_empl***

In [ ]:
%%sql
SELECT
    AVG(number_of_mistakes) AS avg_error_count_per_empl
FROM
    error_count 
limit 5;

## Suspect_list AS A VIEW

In [ ]:
%%sql
DROP VIEW IF EXISTS suspect_list;
CREATE VIEW suspect_list as (
SELECT
    employee_name,
  number_of_mistakes
FROM
    error_count
WHERE 
    number_of_mistakes > (SELECT
                                AVG(number_of_mistakes) AS avg_error_count_per_empl
                            FROM
                                error_count))


In [ ]:
%%sql
SELECT
    *
FROM
    suspect_list;

 ## 7. ISOLATE THEIR RECORDS FROM THE `incorrect_records`
- Look at the Incorrect_records table again and isolate all of the records these four employees gathered.
- Also look at the statements for these records to look for patterns.

## 7.1 Convert the `suspect_list` to a CTE
- use it to filter the records from these four employees
- get the names of the four "suspects", and their mistake count as a result

In [ ]:
%%sql
with suspect_list as (
SELECT
    employee_name,
  number_of_mistakes
FROM
    error_count
WHERE 
    number_of_mistakes > (SELECT
                                AVG(number_of_mistakes) AS avg_error_count_per_empl
                            FROM
                                error_count))
select employee_name
from suspect_list;

## QUERYING THE THREE VIEWS

- Use `Incorrect_records` to find all of the records where the auditor and employee scores don't match.
- USE `error_count` to aggregate the data, and got the number of mistakes each employee made.
- Finally, `suspect_list` retrieves the data of employees who make an above-average number of mistakes.
- Now  filter that Incorrect_records view to identify all of the records associated with the four employees we identified.

## 1. Pull up all of the records where the `employee_name` is in the `suspect list`. HINT: Use `SELECT employee_name FROM suspect_list` as a subquery in WHERE

In [ ]:
%%sql
SELECT
    employee_name,
    location_id,
    statements
FROM
    Incorrect_records
WHERE
    employee_name IN (SELECT 
                          employee_name 
                      FROM suspect_list)
limit 5

In [ ]:
%%sql
WITH error_count AS ( -- This CTE calculates the number of mistakes each employee made
SELECT
    employee_name,
    COUNT(employee_name) AS number_of_mistakes
FROM
incorrect_records
GROUP BY
    employee_name),

suspect_list AS (
SELECT
    employee_name,
    number_of_mistakes
FROM
    error_count
WHERE
    number_of_mistakes > (SELECT AVG(number_of_mistakes) FROM error_count))

SELECT
    employee_name,
    location_id,
    statements
FROM
    incorrect_records
WHERE
    employee_name in (SELECT employee_name FROM suspect_list)
LIMIT 10;

## Corrupt leaders 
- Look at these records: `AkRu04508, AkRu07310,
KiRu29639, AmAm09607`, for example. See how the word "cash" is used a lot in these statements.

In [ ]:
%%sql
SELECT
    employee_name,
    location_id,
   statements
    
FROM
    Incorrect_records

WHERE
    employee_name IN (SELECT 
                          employee_name 
                      FROM suspect_list)
AND location_id	in ('AkRu04508', 'AkRu07310', 'KiRu29639', 'AmAm09607')
limit 5

In [ ]:
%%sql
SELECT
    Incorrect_records.employee_name,
    Incorrect_records.location_id,
    Incorrect_records.statements,
    suspect_list.number_of_mistakes
FROM
    Incorrect_records
JOIN
   suspect_list
ON suspect_list.employee_name=Incorrect_records.employee_name
WHERE
    Incorrect_records.employee_name IN (SELECT 
                          employee_name 
                      FROM suspect_list)
AND Incorrect_records.location_id	in ('AkRu04508', 'AkRu07310', 'KiRu29639', 'AmAm09607')
limit 5

## Filter the records that refer to `"cash"` ---THOSE ALLEGED OF BRIBERY AND CORRUPTION

In [ ]:
%%sql
SELECT
    employee_name,
    location_id,
   statements
    
FROM
    Incorrect_records

WHERE
    employee_name IN (SELECT 
                          employee_name 
                      FROM suspect_list)
    and statements like '%cash%'
limit 10

 ## Check for employees in the `Incorrect_records` table with statements mentioning `"cash"` that are not in our `suspect list.`

In [ ]:
%%sql
SELECT
    employee_name,
    location_id,
   statements
    
FROM
    Incorrect_records

WHERE
    employee_name NOT IN (SELECT 
                          employee_name 
                      FROM suspect_list)
    and statements like '%cash%'
limit 10

***NON***

## ***Sum of the evidence we have for `Zuriel Matembo, Malachi Mavuso, Bello Azibo and Lalitha Kaburi`***

1. They all made more mistakes than their peers on average.
2. They all have incriminating statements made against them, and only them.